platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,826 @@
+2. TssMessage - the raw bytes of the TssMessage
+
+The value associated with a TSS_VOTE key is a bit vector with the following interpretation
+
+* The order of bits from the least significant bit to the most significant bit corresponds to the numeric order of
+  share ids from least to greatest.
+* If a bit is set to 1, then the TssMessage for the corresponding ShareId was valid and contributed to a successful
+  reconstruction of the ledger id.
+* If a bit is set to 0, then the TssMessage for the corresponding ShareId was either invalid, was not received, or
+  was not used in the reconstruction of the ledger id.
+* If the sum of the bits is greater than the threshold, then the vote is a yes vote in favor of adopting the
+  candidate roster.
+
+Lifecycle Invariants
+
+1. Inserts should only be made for candidate rosters that are going through the keying process.
+2. When a roster is no longer stored in the state, the corresponding key material should be removed.
+3. Every roster that has become active after the TSS bootstrap process must have sufficient key material to be
+   able to recover the ledger id.
+
+TODO: Protobuf of data
+
+##### Partial Ledger Signatures
+
+For a configured number of rounds, share signatures on a message are stored in the state until a threshold number of
+them have been collected to produce the ledger id. Each time a signature on a message is added to the collection,
+the round counter resets. Once the ledger id is produced, it is stored with the other signatures and no new
+signatures are added to the collection. Once a configured number of rounds have passed after the last signature
+added to a collection for a message, the signatures are removed from the state.
+
+The partial ledger signatures are stored in a Map Merkle Leaf where the keys are the message hash and the values are
+a tuple (List<ShareSignature>, LedgerSignature, lastRoundUpdated).
+
+TODO: Protobuf of data
+
+### Core Behaviors
+
+The new behavior is related to the bootstrap process of creating the ledger id, transferring the ability to sign
+messages from one roster to the next roster, and creating ledger signatures.
+
+#### TSS Bootstrap for New Networks
+
+At the start of a new network there will be a new pre-genesis phase to generate the key material for the genesis
+roster and then restart the network from round 0 with the new key material and ledger id.
+
+The following startup sequence is modified from existing practices.
+
+1. Inversion of Control - The app hands a genesis state, genesis roster, and private keys to the platform.
+2. The platform copies the genesis state and starts gossiping with peers without accepting user transactions.
+3. The platform automatically begins to key the genesis roster with a ledger id.
+4. Once the ledger id is created and a threshold number of nodes have voted to being ready to restart, the platform
+   updates the state with the key material and ledger id and restarts from round 0.
```

Comment:
> updates the state

It's unclear to me what "the state" is referring to here.

The described sequence of events seems to have the platform in the driver's seat. Am I understanding your intention correctly?

If so, I'm wondering if it's the correct strategy. Here is an alternate sequence I'm thinking about:
1. Platform is started, and given a genesis state (which, of course, lacks key material)
2. Platform does the genesis keying, and communicates in some way that it is "done"
3. The platform is restarted by an external source, and given a different genesis state after restarting (this time, including the key material, which the external source has extracted from the pre-genesis state)

With this sequence, the platform isn't responsible for copying data from here to there. It just does the keying, puts the key material in the state, and that's it.

Reply:
Your alternative requires this additional behavior: 
1. communicating to the App that it is done with pre-genesis.   (The app now has to be aware of the more complicated platform life-cycle so it can restart the platform.) 
2.a. The key material has to be handed to the app and the app has to figure out how to add the key-material to the state (something that the platform should do.) 
2.b.  Alternatively the platform hands back the modified genesis state to the app. 
3. The app restarts the platform and hands back the genesis state with the key material added to it. 

2.a.  is way too much of a burden on the app and its something the platform is already doing. 
2.b. makes little sense since the app is going to hand the state it just got right back to the platform. 

The platform just needs a small bit of logic at the beginning to restart itself and allow the app to be oblivious that the platform restart is even happening.  

I'm not convinced burdening the app with this responsibility is the right solution.